File: src_py/playwright_search/browser_manager.py
import asyncio
import json
import os
import random
from datetime import datetime
from typing import Optional, Tuple

from playwright.async_api import Browser, BrowserContext, async_playwright, Playwright, devices
from playwright_stealth import stealth_async

from .logger import logger
from .types import CommandOptions


class BrowserManager:
    """负责管理浏览器实例、上下文和反爬虫策略"""

    def __init__(self, options: CommandOptions):
        self.options = options
        self.state_file = options.state_file or "./browser-state.json"
        self.fingerprint_file = self.state_file.replace(".json", "-fingerprint.json")
        self.playwright: Optional[Playwright] = None
        self.browser: Optional[Browser] = None
        self.context: Optional[BrowserContext] = None

    async def launch_browser(self) -> BrowserContext:
        """启动浏览器，应用反爬虫插件，并创建浏览器上下文"""
        logger.info("正在启动浏览器...")
        self.playwright = await async_playwright().start()

        browser_args = [
            "--no-sandbox",
            "--disable-setuid-sandbox",
            "--disable-dev-shm-usage",
            "--disable-accelerated-2d-canvas",
            "--no-first-run",
            "--no-zygote",
            "--disable-gpu",
        ]
        
        # 启动 Chromium 浏览器
        self.browser = await self.playwright.chromium.launch(
            headless=self.options.headless if self.options.headless is not None else True,
            args=browser_args
        )

        logger.info("浏览器启动成功，正在创建浏览器上下文...")

        # 加载或创建指纹
        storage_state, saved_state = self._load_saved_state()
        fingerprint = saved_state.get("fingerprint") or self._get_host_machine_config()

        context_options = {
            "locale": fingerprint.get("locale"),
            "timezone_id": fingerprint.get("timezoneId"),
            "color_scheme": fingerprint.get("colorScheme"),
            "reduced_motion": fingerprint.get("reducedMotion"),
            "forced_colors": fingerprint.get("forcedColors"),
            "user_agent": fingerprint.get("userAgent"),
        }
        
        # 应用设备模拟
        if fingerprint.get("deviceName"):
            device_name = fingerprint["deviceName"]
            if device_name in devices:
                context_options.update(devices[device_name])

        if storage_state:
            context_options["storage_state"] = storage_state

        self.context = await self.browser.new_context(**context_options)
        
        # 应用 stealth 插件
        logger.info("正在应用 Stealth 插件以增强反爬虫能力...")
        await stealth_async(self.context)
        logger.info("Stealth 插件应用成功")

        # 设置默认超时
        self.context.set_default_timeout(self.options.timeout or 60000)

        logger.info({"fingerprint": fingerprint}, "浏览器上下文创建成功")
        return self.context

    def _load_saved_state(self) -> Tuple[Optional[str], dict]:
        """加载保存的浏览器状态和指纹"""
        storage_state = None
        saved_state = {}

        if os.path.exists(self.state_file):
            logger.info(f"发现浏览器状态文件: {self.state_file}, 将加载状态以避免反机器人检测。")
            storage_state = self.state_file
            if os.path.exists(self.fingerprint_file):
                try:
                    with open(self.fingerprint_file, "r", encoding="utf-8") as f:
                        saved_state = json.load(f)
                    logger.info("已加载保存的浏览器指纹。")
                except Exception as e:
                    logger.warning(f"无法加载指纹文件: {e}, 将创建新指纹。")
        else:
            logger.info("未找到浏览器状态文件，将创建新的会话和指纹。")
            
        return storage_state, saved_state

    def _get_host_machine_config(self) -> dict:
        """获取主机配置作为指纹"""
        system_locale = self.options.locale or os.getenv("LANG", "zh-CN")
        hour = datetime.now().hour
        color_scheme = "dark" if hour >= 18 or hour <= 6 else "light"

        device_list = [
            "Desktop Chrome",
            "Desktop Edge",
            "Desktop Firefox",
        ]
        
        device_name = random.choice(device_list)
        user_agent = devices[device_name]['user_agent']

        return {
            "deviceName": device_name,
            "locale": system_locale,
            "timezoneId": random.choice(["America/New_York", "Europe/London", "Asia/Shanghai"]),
            "colorScheme": color_scheme,
            "reducedMotion": "no-preference",
            "forcedColors": "none",
            "userAgent": user_agent,
        }

    async def close(self):
        """保存状态并关闭浏览器"""
        if self.context and not (self.options.no_save_state):
            logger.info("正在保存浏览器状态和指纹...")
            try:
                state_dir = os.path.dirname(self.state_file)
                if not os.path.exists(state_dir):
                    os.makedirs(state_dir)
                
                # 保存状态
                await self.context.storage_state(path=self.state_file)

                # 保存指纹
                fingerprint = self._get_host_machine_config() # 使用最新的配置
                with open(self.fingerprint_file, "w", encoding="utf-8") as f:
                    json.dump({"fingerprint": fingerprint}, f, indent=2)

                logger.info("浏览器状态和指纹保存成功。")
            except Exception as e:
                logger.warning(f"保存浏览器状态失败: {e}")

        if self.browser:
            await self.browser.close()
            logger.info("浏览器已关闭。")
        if self.playwright:
            await self.playwright.stop()

File: src/engines/browser-manager.ts
import { chromium, Browser, BrowserContext, devices } from "playwright";
import * as fs from "fs";
import * as path from "path";
import * as os from "os";
import { CommandOptions } from "../types.js";
import logger from "../logger.js";
import { ChromiumBrowserManager } from "./chromium-manager.js";

// 指纹配置接口
export interface FingerprintConfig {
  deviceName: string;
  locale: string;
  timezoneId: string;
  colorScheme: "dark" | "light";
  reducedMotion: "reduce" | "no-preference";
  forcedColors: "active" | "none";
  viewport?: { width: number; height: number };
  userAgent?: string;
}

// 新：每个引擎的独立状态
export interface EngineState {
  fingerprint?: FingerprintConfig;
  proxy?: string;
  googleDomain?: string; // 特定于Google的配置
}

// 更新：保存的状态文件接口，现在是一个映射
export type SavedState = Record<string, EngineState>;

// 浏览器管理器抽象类
export abstract class BaseBrowserManager {
  protected options: CommandOptions;
  protected stateFile: string;
  protected fingerprintFile: string;

  constructor(options: CommandOptions = {}) {
    this.options = options;

    if (options.stateFile) {
      this.stateFile = options.stateFile;
    } else {
      let stateDir: string;
      const localStateDir = path.resolve(".playwright-search");
      const homeStateDir = path.join(os.homedir(), ".playwright-search");

      if (fs.existsSync(localStateDir)) {
        stateDir = localStateDir;
      } else {
        stateDir = homeStateDir;
      }
      // Ensure the state directory exists
      if (!fs.existsSync(stateDir)) {
        fs.mkdirSync(stateDir, { recursive: true });
      }
      this.stateFile = path.join(stateDir, "browser-state.json");
    }

    this.fingerprintFile = this.stateFile.replace(
      ".json",
      "-fingerprint.json"
    );

    
    logger.info(`加载配置文件: ${this.fingerprintFile}`);
  }

  // 抽象方法：子类必须实现
  abstract createBrowser(): Promise<Browser>;
  abstract createContext(
    browser: Browser,
    engineState: EngineState,
    proxy?: string
  ): Promise<BrowserContext>;

  // 更新：加载特定引擎的状态
  public loadEngineState(engineId: string): EngineState {
    const allStates = this.loadSavedStatesFromFile();
    return allStates[engineId] || {};
  }

  // 私有方法：从文件加载所有状态
  private loadSavedStatesFromFile(): SavedState {
    let savedState: SavedState = {};

    if (fs.existsSync(this.fingerprintFile)) {
      try {
        const fingerprintData = fs.readFileSync(this.fingerprintFile, "utf8");
        savedState = JSON.parse(fingerprintData);
        logger.info("已加载所有引擎的浏览器指纹和代理配置");
      } catch (e) {
        logger.warn(
          { error: e },
          "无法加载指紋或代理配置文件，将创建新的"
        );
      }
    }
    return savedState;
  }

  // 更新：保存状态和指纹
  public async saveStateAndFingerprint(
    context: BrowserContext,
    engineId: string,
    engineState: EngineState,
    noSaveState: boolean = false
  ): Promise<void> {
    if (noSaveState) return;

    try {
      const stateDir = path.dirname(this.stateFile);
      if (!fs.existsSync(stateDir)) {
        fs.mkdirSync(stateDir, { recursive: true });
      }

      // 保存浏览器上下文（cookies, local storage等）
      await context.storageState({ path: this.stateFile });

      // 加载当前所有引擎的状态，以避免覆盖
      const currentState = this.loadSavedStatesFromFile();
      
      // 更新或添加当前引擎的状态
      currentState[engineId] = {
        ...currentState[engineId], // 保留该引擎可能存在的旧属性
        ...engineState,
      };

      // 将更新后的完整状态写回文件
      fs.writeFileSync(
        this.fingerprintFile,
        JSON.stringify(currentState, null, 2),
        "utf8"
      );

      logger.info(`已为引擎 '${engineId}' 保存浏览器状态、指纹和代理配置`);
    } catch (e) {
      logger.warn({ error: e }, "保存浏览器状态/指纹/代理失败");
    }
  }

  // 通用方法：获取随机设备配置
  protected getRandomDeviceConfig(): [string, any] {
    // 筛选出所有桌面设备
    const desktopDevices = Object.keys(devices).filter(
      (name) => !devices[name].isMobile
    ).filter(
      (name) => devices[name].userAgent.indexOf("Chrome") !== -1
    );

    // 从桌面设备中随机选择一个
    const randomDeviceName =
      desktopDevices[Math.floor(Math.random() * desktopDevices.length)];
    const device = devices[randomDeviceName];

    // 强制设置720p分辨率
    device.viewport = { width: 1200, height: 768 };

    return [randomDeviceName, device];
  }

  // 通用方法：获取随机时区
  protected getRandomTimezone(): string {
    const timezoneList = [
      "Asia/Shanghai"
    ];

    return timezoneList[Math.floor(Math.random() * timezoneList.length)];
  }

  // 通用方法：获取宿主机器的实际配置
  public getHostMachineConfig(userLocale?: string): FingerprintConfig {
    const systemLocale = userLocale || process.env.LANG || "zh-CN";
    const timezoneId = this.getRandomTimezone();
    const hour = new Date().getHours();
    const colorScheme: "dark" | "light" =
      hour >= 18 || hour <= 6 ? "dark" : "light";

    const [deviceName, device] = this.getRandomDeviceConfig();

    return {
      deviceName,
      locale: systemLocale,
      timezoneId,
      colorScheme,
      reducedMotion: "no-preference",
      forcedColors: "none",
      viewport: device.viewport,
      userAgent: device.userAgent,
    };
  }

  // 通用方法：规范化 headless 配置
  protected coerceHeadless(value: unknown): boolean {
    if (value === false) return false;
    if (typeof value === "string") {
      const v = value.toLowerCase();
      if (v === "false" || v === "0" || v === "no") return false;
    }
    return true;
  }

  // 通用方法：获取随机延迟时间
  protected getRandomDelay(min: number, max: number): number {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }
  // 通用方法：解析代理配置
  protected parseProxyConfig(proxyUrl: string): any {
    try {
      const u = new URL(proxyUrl);
      const server = `${u.protocol}//${u.hostname}${
        u.port ? ":" + u.port : ""
      }`;
      const cfg: { server: string; username?: string; password?: string } = {
        server,
      };
      if (u.username) cfg.username = decodeURIComponent(u.username);
      if (u.password) cfg.password = decodeURIComponent(u.password);
      return cfg;
    } catch (e) {
      logger.warn({ proxy: proxyUrl }, "代理URL解析失败");
      return { server: proxyUrl };
    }
  }
}
